File: tasks/translation/output/Qwen/Qwen2.5-Coder-32B-Instruct-AWQ/codenet/direct/Java/Python/s514557145_commented.py
MOD1 = 1000000007
MOD2 = 998244353
tenmod = []
inv9 = pow(9, -1, MOD2)

class S:
    E = None
    def __init__(self, sum, l, r):
        self.sum = sum
        self.l = l
        self.r = r

    @staticmethod
    def op(s1, s2):
        sum = (s1.sum + s2.sum) % MOD2
        return S(sum, min(s1.l, s2.l), max(s1.r, s2.r))

    @staticmethod
    def map(f, s):
        c = (tenmod[f.N - s.l] - tenmod[f.N - s.r - 1]) % MOD2
        return S(((f.D * c) % MOD2 * inv9) % MOD2, s.l, s.r)

S.E = S(0, float('inf') // 2, -float('inf') // 2)

class F:
    I = None
    def __init__(self, D, N):
        self.D = D
        self.N = N

    @staticmethod
    def composite(f, g):
        return F(f.D, f.N)

F.I = F(0, 0)

class LazySegTree:
    def __init__(self, n, op, e, mapping, composition, id):
        self.MAX = n
        k = 1
        while k < n:
            k <<= 1
        self.N = k
        self.Log = k.bit_length() - 1
        self.Op = op
        self.E = e
        self.Mapping = mapping
        self.Composition = composition
        self.Id = id
        self.Dat = [e] * (2 * self.N)
        self.Laz = [id] * self.N

    def build(self, dat):
        l = len(dat)
        self.Dat[self.N:self.N + l] = dat
        for i in range(self.N - 1, 0, -1):
            self.Dat[i] = self.Op(self.Dat[2 * i], self.Dat[2 * i + 1])

    def push(self, k):
        if self.Laz[k] == self.Id:
            return
        self.Dat[2 * k] = self.Mapping(self.Laz[k], self.Dat[2 * k])
        self.Dat[2 * k + 1] = self.Mapping(self.Laz[k], self.Dat[2 * k + 1])
        if 2 * k < self.N:
            self.Laz[2 * k] = self.Composition(self.Laz[k], self.Laz[2 * k])
            self.Laz[2 * k + 1] = self.Composition(self.Laz[k], self.Laz[2 * k + 1])
        self.Laz[k] = self.Id

    def push_to(self, k):
        for i in range(self.Log, 0, -1):
            self.push(k >> i)

    def push_to_range(self, lk, rk):
        for i in range(self.Log, 0, -1):
            if (lk >> i) << i != lk:
                self.push(lk >> i)
            if (rk >> i) << i != rk:
                self.push((rk - 1) >> i)

    def update_from(self, k):
        k >>= 1
        while k > 0:
            self.Dat[k] = self.Op(self.Dat[2 * k], self.Dat[2 * k + 1])
            k >>= 1

    def update_from_range(self, lk, rk):
        for i in range(1, self.Log + 1):
            if (lk >> i) << i != lk:
                self.Dat[lk >> i] = self.Op(self.Dat[2 * (lk >> i)], self.Dat[2 * (lk >> i) + 1])
            if (rk >> i) << i != rk:
                self.Dat[(rk - 1) >> i] = self.Op(self.Dat[2 * ((rk - 1) >> i)], self.Dat[2 * ((rk - 1) >> i) + 1])

    def set(self, p, x):
        if p < 0 or p >= self.MAX:
            raise IndexError(f"Index {p} is not in [0, {self.MAX}).")
        p += self.N
        self.push_to(p)
        self.Dat[p] = x
        self.update_from(p)

    def get(self, p):
        if p < 0 or p >= self.MAX:
            raise IndexError(f"Index {p} is not in [0, {self.MAX}).")
        p += self.N
        self.push_to(p)
        return self.Dat[p]

    def all_prod(self):
        return self.Dat[1]

    def apply(self, p, f):
        if p < 0 or p >= self.MAX:
            raise IndexError(f"Index {p} is not in [0, {self.MAX}).")
        p += self.N
        self.push_to(p)
        self.Dat[p] = self.Mapping(f, self.Dat[p])
        self.update_from(p)

    def apply_range(self, l, r, f):
        if l > r:
            raise ValueError(f"Invalid range: [{l}, {r}).")
        if l < 0 or l > self.MAX or r < 0 or r > self.MAX:
            raise IndexError(f"Range [{l}, {r}) is not in [0, {self.MAX}].")
        if l == r:
            return
        l += self.N
        r += self.N
        self.push_to_range(l, r)
        lk, rk = l, r
        while l < r:
            if l & 1:
                self.Dat[l] = self.Mapping(f, self.Dat[l])
                if l < self.N:
                    self.Laz[l] = self.Composition(f, self.Laz[l])
                l += 1
            if r & 1:
                r -= 1
                self.Dat[r] = self.Mapping(f, self.Dat[r])
                if r < self.N:
                    self.Laz[r] = self.Composition(f, self.Laz[r])
            l >>= 1
            r >>= 1
        self.update_from_range(lk, rk)

    def max_right(self, l, g):
        if l < 0 or l > self.MAX:
            raise IndexError(f"Index {l} is not in [0, {self.MAX}].")
        if not g(self.E):
            raise ValueError("Identity element must satisfy the condition.")
        if l == self.MAX:
            return self.MAX
        l += self.N
        self.push_to(l)
        sum = self.E
        while True:
            l >>= (l & -l).bit_length() - 1
            if not g(self.Op(sum, self.Dat[l])):
                while l < self.N:
                    self.push(l)
                    l <<= 1
                    if g(self.Op(sum, self.Dat[l])):
                        sum = self.Op(sum, self.Dat[l])
                        l += 1
                return l - self.N
            sum = self.Op(sum, self.Dat[l])
            l += 1
            if (l & -l) == l:
                break
        return self.MAX

    def min_left(self, r, g):
        if r < 0 or r > self.MAX:
            raise IndexError(f"Index {r} is not in [0, {self.MAX}].")
        if not g(self.E):
            raise ValueError("Identity element must satisfy the condition.")
        if r == 0:
            return 0
        r += self.N
        self.push_to(r - 1)
        sum = self.E
        while True:
            r -= 1
            while r > 1 and r & 1:
                r >>= 1
            if not g(self.Op(self.Dat[r], sum)):
                while r < self.N:
                    self.push(r)
                    r = r << 1 | 1
                    if g(self.Op(self.Dat[r], sum)):
                        sum = self.Op(self.Dat[r], sum)
                        r -= 1
                return r + 1 - self.N
            sum = self.Op(self.Dat[r], sum)
            if (r & -r) == r:
                break
        return 0

import sys
import io

class InputReader:
    def __init__(self, input_stream):
        self.input_stream = input_stream
        self.buffer = io.BytesIO()
        self.input_stream.seek(0, 2)
        self.input_stream.seek(0)
        self.buffer.write(self.input_stream.read())
        self.buffer.seek(0)
        self.input_stream.close()
        self.input_stream = self.buffer
        self.current_line = []
        self.current_index = 0

    def read(self, size=-1):
        if size == -1:
            size = sys.maxsize
        while len(self.current_line) <= self.current_index:
            line = self.input_stream.readline()
            if not line:
                break
            self.current_line.extend(line.split())
        result = self.current_line[self.current_index:self.current_index + size]
        self.current_index += size
        return result

    def next(self):
        return self.read(1)[0]

    def next_int(self):
        return int(self.next())

    def next_long(self):
        return int(self.next())

    def next_double(self):
        return float(self.next())

    def next_int_array(self, n):
        return list(map(int, self.read(n)))

    def next_long_array(self, n):
        return list(map(int, self.read(n)))

    def next_char_map(self, n, m):
        return [list(self.next()) for _ in range(n)]

import sys

def main():
    import sys
    import io
    input = sys.stdin.read
    data = input().split()
    index = 0

    def next():
        nonlocal index
        result = data[index]
        index += 1
        return result

    def next_int():
        return int(next())

    def next_long():
        return int(next())

    def next_double():
        return float(next())

    def next_int_array(n):
        return list(map(int, next().split()))

    def next_long_array(n):
        return list(map(int, next().split()))

    def next_char_map(n, m):
        return [list(next()) for _ in range(n)]

    N = next_int()
    Q = next_int()
    global tenmod
    tenmod = [1] * (N + 1)
    for j in range(1, N + 1):
        tenmod[j] = (tenmod[j - 1] * 10) % MOD2
    dat = [S(tenmod[N - i - 1], i, i) for i in range(N)]
    lazySegTree = LazySegTree(N, S.op, S.E, S.map, F.composite, F.I)
    for j in range(Q):
        l = next_int() - 1
        r = next_int() - 1
        D = next_long()
        lazySegTree.apply_range(l, r + 1, F(D, N))
        print(lazySegTree.all_prod().sum)

if __name__ == "__main__":
    main()
# End of Code

File: tasks/translation/output/Qwen/Qwen2.5-Coder-32B-Instruct-AWQ/codenet/direct/Java/Go/s514557145_commented.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const (
	MOD1 = 1000000007
	MOD2 = 998244353
)

var tenmod []int64

func modInv(x int64) int64 {
	return modPow(x, MOD2-2)
}

func modPow(x, y int64) int64 {
	z := int64(1)
	for y > 0 {
		if y%2 == 0 {
			x = (x * x) % MOD2
			y /= 2
		} else {
			z = (z * x) % MOD2
			y--
		}
	}
	return z
}

type S struct {
	sum int64
	l   int
	r   int
}

func (s S) op(s2 S) S {
	sum := s.sum + s2.sum
	if sum >= MOD2 {
		sum -= MOD2
	}
	return S{sum, min(s.l, s2.l), max(s.r, s2.r)}
}

func (s S) mapF(f F) S {
	c := (tenmod[f.N-s.l] - tenmod[f.N-s.r-1])
	if c < 0 {
		c += MOD2
	}
	inv9 := modInv(9)
	return S{((f.D * c) % MOD2 * inv9) % MOD2, s.l, s.r}
}

type F struct {
	D int64
	N int
}

func (f F) composite(f2 F) F {
	return F{f.D, f.N}
}

var (
	S_E = S{0, (1 << 31) / 2, -(1 << 31) / 2}
	F_I = F{0, 0}
)

type LazySegTree struct {
	N           int
	Log         int
	Op          func(S, S) S
	E           S
	Mapping     func(F, S) S
	Composition func(F, F) F
	Id          F
	Dat         []S
	Laz         []F
}

func NewLazySegTree(n int, op func(S, S) S, e S, mapping func(F, S) S, composition func(F, F) F, id F) *LazySegTree {
	k := 1
	for k < n {
		k <<= 1
	}
	lst := &LazySegTree{
		N:           k,
		Log:         32 - bitsLeadingZeros(uint(k)),
		Op:          op,
		E:           e,
		Mapping:     mapping,
		Composition: composition,
		Id:          id,
		Dat:         make([]S, k*2),
		Laz:         make([]F, k),
	}
	for i := range lst.Dat {
		lst.Dat[i] = e
	}
	for i := range lst.Laz {
		lst.Laz[i] = id
	}
	return lst
}

func (lst *LazySegTree) build(dat []S) {
	l := len(dat)
	copy(lst.Dat[lst.N:], dat)
	for i := lst.N - 1; i > 0; i-- {
		lst.Dat[i] = lst.Op(lst.Dat[i<<1], lst.Dat[i<<1|1])
	}
}

func (lst *LazySegTree) push(k int) {
	if lst.Laz[k] == F_I {
		return
	}
	lk, rk := k<<1, k<<1|1
	lst.Dat[lk] = lst.Mapping(lst.Laz[k], lst.Dat[lk])
	lst.Dat[rk] = lst.Mapping(lst.Laz[k], lst.Dat[rk])
	if lk < lst.N {
		lst.Laz[lk] = lst.Composition(lst.Laz[k], lst.Laz[lk])
	}
	if rk < lst.N {
		lst.Laz[rk] = lst.Composition(lst.Laz[k], lst.Laz[rk])
	}
	lst.Laz[k] = F_I
}

func (lst *LazySegTree) pushTo(k int) {
	for i := lst.Log; i > 0; i-- {
		lst.push(k >> i)
	}
}

func (lst *LazySegTree) pushTo(lk, rk int) {
	for i := lst.Log; i > 0; i-- {
		if ((lk >> i) << i) != lk {
			lst.push(lk >> i)
		}
		if ((rk >> i) << i) != rk {
			lst.push(rk >> i)
		}
	}
}

func (lst *LazySegTree) updateFrom(k int) {
	k >>= 1
	for k > 0 {
		lst.Dat[k] = lst.Op(lst.Dat[k<<1], lst.Dat[k<<1|1])
		k >>= 1
	}
}

func (lst *LazySegTree) updateFrom(lk, rk int) {
	for i := 1; i <= lst.Log; i++ {
		if ((lk >> i) << i) != lk {
			lki := lk >> i
			lst.Dat[lki] = lst.Op(lst.Dat[lki<<1], lst.Dat[lki<<1|1])
		}
		if ((rk >> i) << i) != rk {
			rki := (rk - 1) >> i
			lst.Dat[rki] = lst.Op(lst.Dat[rki<<1], lst.Dat[rki<<1|1])
		}
	}
}

func (lst *LazySegTree) set(p int, x S) {
	if p < 0 || p >= lst.N {
		panic("Index out of range")
	}
	p += lst.N
	lst.pushTo(p)
	lst.Dat[p] = x
	lst.updateFrom(p)
}

func (lst *LazySegTree) get(p int) S {
	if p < 0 || p >= lst.N {
		panic("Index out of range")
	}
	p += lst.N
	lst.pushTo(p)
	return lst.Dat[p]
}

func (lst *LazySegTree) allProd() S {
	return lst.Dat[1]
}

func (lst *LazySegTree) apply(p int, f F) {
	if p < 0 || p >= lst.N {
		panic("Index out of range")
	}
	p += lst.N
	lst.pushTo(p)
	lst.Dat[p] = lst.Mapping(f, lst.Dat[p])
	lst.updateFrom(p)
}

func (lst *LazySegTree) applyRange(l, r int, f F) {
	if l > r {
		panic("Invalid range")
	}
	if l < 0 || r > lst.N {
		panic("Index out of range")
	}
	if l == r {
		return
	}
	l += lst.N
	r += lst.N
	lst.pushTo(l, r)
	for l < r {
		if l&1 == 1 {
			lst.Dat[l] = lst.Mapping(f, lst.Dat[l])
			if l < lst.N {
				lst.Laz[l] = lst.Composition(f, lst.Laz[l])
			}
			l++
		}
		if r&1 == 1 {
			r--
			lst.Dat[r] = lst.Mapping(f, lst.Dat[r])
			if r < lst.N {
				lst.Laz[r] = lst.Composition(f, lst.Laz[r])
			}
		}
		l >>= 1
		r >>= 1
	}
	lst.updateFrom(l, r)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func bitsLeadingZeros(x uint) int {
	if x == 0 {
		return 32
	}
	return 31 - bitsTrailingZeros(x)
}

func bitsTrailingZeros(x uint) int {
	if x == 0 {
		return 32
	}
	var n int
	for x&1 == 0 {
		n++
		x >>= 1
	}
	return n
}

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	N := readInt(in)
	Q := readInt(in)

	tenmod = make([]int64, N+1)
	tenmod[0] = 1
	for j := 1; j <= N; j++ {
		tenmod[j] = (tenmod[j-1] * 10) % MOD2
	}

	dat := make([]S, N)
	for i := 0; i < N; i++ {
		dat[i] = S{tenmod[N-i-1], i, i}
	}

	lazySegTree := NewLazySegTree(N, S.op, S_E, S.mapF, F.composite, F_I)
	lazySegTree.build(dat)

	for j := 0; j < Q; j++ {
		l := readInt(in) - 1
		r := readInt(in) - 1
		D := readInt64(in)
		lazySegTree.applyRange(l, r+1, F{D, N})
		fmt.Fprintln(out, lazySegTree.allProd().sum)
	}
}

func readInt(in *bufio.Reader) int {
	line, _ := in.ReadString('\n')
	line = strings.TrimSpace(line)
	v, _ := strconv.Atoi(line)
	return v
}

func readInt64(in *bufio.Reader) int64 {
	line, _ := in.ReadString('\n')
	line = strings.TrimSpace(line)
	v, _ := strconv.ParseInt(line, 10, 64)
	return v
}

// End of Code
